### ISSUE TITLE
Design proposal for RTCIceTransport

### ISSUE THREAD

**Issue by trivigy** [pions/webrtc#83]
For a little white I have been brainstorming on how to best enhance the ICE protocol and finally after conversations with @Sean-Der decided to put down the information I found in this proposal. This change is primarily only for enhancing internal readability and does not change a single thing in terms of the public API. In this proposal I am not going to cover in depth how the current design of the IceAgent is structures In fact for the most part how it operates is really not negotiable since it needs to comply with [RFC5245] (https://tools.ietf.org/html/rfc5245). 
>If you are interested please follow the code [HERE](https://github.com/pions/webrtc/blob/master/pkg/ice/agent.go#L44). Current implementation uses this object in two places. During initialization [HERE](https://github.com/pions/webrtc/blob/master/internal/network/manager.go#L66-L81) and when starting up the gathering process [HERE](https://github.com/pions/webrtc/blob/master/signaling.go#L145).

What I will mention however is why in my opinion this design is hard to read and how to package as its own module:
1. The code is initialized during application initialization but, in fact nothing within the module is actually used during that intialization nor is there any initialization specific information covered in the ICE RFC. This fact seems to have been actually addressed and extended using [RTCPeerConnection Interface Extensions] (https://www.w3.org/TR/webrtc/#rtcpeerconnection-interface-extensions) which talks about an object called `RTCIceTransport` whose suppose is to be an interface into the ICE process.
2. The whole process of gathering is kicked off with a `start()` function which does not convey the needed information that this is a gathering of candidates process.
3. Local candidate gathering is happening outside of the `IceAgent`. This makes things hard for isolation purposes.
4. `IceAgent` shares objects with the rest of the webrtc. This one makes it especially hard to make ICE its own module. In fact after reading the entire ICE RFC there is absolutely no mentioning of webrtc state management which raises the question of why ICE shares objects with webrtc at all.

**Solution**:
> Note this is not my solution an I am simply adopting the extended variation of the webrtc spec.

After a few days of investigation and trying to figure out how exactly the `RTCIceTransport` plays a role relative to the `IceAgent` that we created I came to the conclusion that they are one and the same. (In a way). WebRTC spec talks a lot about Ice Transports. In fact a lot of the state management of webrtc is done by checking the states and existence of these ice transports. For a little while I could not understand what they are but I after following up on other webrtc library in another language (python) I found one that clearly explains it. [HERE] (https://github.com/jlaine/aiortc/blob/master/aiortc/rtcpeerconnection.py#L108) you can see that the library initializes an empty set of `iceTransports`. Then during `addTrack` these `iceTransports` are initialized and places inside of this set ([1](https://github.com/jlaine/aiortc/blob/master/aiortc/rtcpeerconnection.py#L329)->[2](https://github.com/jlaine/aiortc/blob/master/aiortc/rtcpeerconnection.py#L561)->[3](https://github.com/jlaine/aiortc/blob/master/aiortc/rtcpeerconnection.py#L484-L488)). In fact they are first wrapped inside of a dtls object. Then during the `setLocalDescription` function the gathering process is kicked off [HERE](https://github.com/jlaine/aiortc/blob/master/aiortc/rtcpeerconnection.py#L333) which  does the whole ice compliant process.

**Here is the proposal** (_pseudocode_):
```
class IceAgent:
    AddRemoteCandidate()
    Gather()
    Connect()
    SetSelectedPair()
```
> For the most part it will be similar to this design: https://aioice.readthedocs.io/en/stable/api.html

The initialization process is going to be during `addTrack` or `createDataChannel` depending on whether this is a data transfer or audio/video. Just like the python library we'll wrap the newly create `RTCIceTransport` with a `RTCDtlsTransport` object. And finally we'll kick off the complete gathering and connectivity processes during `setLocalDescription`. Just like the WebRTC RFC instructs to do.

The design inspiration credits go to: https://github.com/jlaine.
Just want to give credit where it is due.

**Comment by Sean-Der** [comment]
This is fantastic! Thank you so much for doing this, in the future this should be the standard before making any changes, this just makes everything better long term.

* For the IceAgent constructor we pass STUN/TURN URLs. It then maintains them and emits up to the RTCPeerConnection when they are ready. This makes sense to me (and matches the RFC) thanks for doing all the research!

* `IceAgent.Gather` is this async? Would be nice if `trickle-ice` landed with this, so you only have to do this all once. This is just fired when the RTCPeerConnection is built (so we start contacting STUN/TURN before setLocalDescription/setRemoteDescription) 

* Maybe change `Connect` to `Start` to match aioice and the intent a little better (since we aren't really connecting, but kicking off a complicated stateful process)  

* I find `RTCDtlsTransport` confusing. Especially since it is only used for SCTP traffic, and it shared across all candidates. Currently we share one DTLS state between all ports, and candidates may switch but we always use the same DTLS state. Right now a candidate sends traffic, and just conditionally chooses to encrypt things using DTLS (am I sending SRTP or SCTP).

What does the call-stack look like for sending an RTP/SCTP packet then? Do we actually send via these transports, or just query things from them?

`RTCDataChannel.Send() -> iceAgent.SelectedPair() -> selectedPair.SendRTP() -> encrypts+send `
`r.networkManager.SendRTP -> iceAgent.SelectedPair() -> selectedPair.SendSRTP -> encrypts + send`

* I don't know how much value you will get from unique packages, I think the overhead of versioning/releases is more then you expect. If someone is looking for these packages I would love to build it for them, but I am a little worried spending time on things that no one uses.
-----
I am 100% for making a better ICEAgent/keeping more state in there!
Would like a little more explanation about DTLSTransport.

Personally I don't think ORTC/WebRTC is perfect, it is just an arbitrary design and at times seems cryptic so I think it is ok to deviate, if we get wins in maintainability/onboard new devs.

**Comment by trivigy** [comment]
@Sean-Der So yeah... I wrote an answer for every point you mentioned. Sorry that it is a long explanation.

> For the IceAgent constructor we pass STUN/TURN URLs. It then maintains them and emits up to the RTCPeerConnection when they are ready. This makes sense to me (and matches the RFC) thanks for doing all the research!

Yes. That exactly how it is going to work.

> IceAgent.Gather is this async? Would be nice if trickle-ice landed with this, so you only have to do this all once. This is just fired when the RTCPeerConnection is built (so we start contacting STUN/TURN before setLocalDescription/setRemoteDescription)

A few things to mention here. As for the async capability no. At least not at first. In my personal opinion we should use goroutines in general because **a**. it is SUPER golang. **b**. It is in line with RFC design. **c**. Without this we, already, have issues with parallelism. My product has a bunch of concurrency and this library as is currently not going to be usable by me for a little until I can test that there are no race conditions created. But really it is a big issue that will keep coming up more and more as the library becomes more stable. I am happy to help there. That is the stuff I like the most personally. 

As for contacting before `setLocalDescription`/`setRemoteDescription`. According to the JSEP RFC [HERE](https://tools.ietf.org/html/draft-ietf-rtcweb-jsep-24#section-7.1) gathering only starts during `setLocalDescription` and not before that. In fact that is also confirmed by the ICE RFC [HERE](https://tools.ietf.org/html/rfc5245#section-4.1.1) when it says: "An agent gathers candidates when it believes that communication is imminent." I am going to keep emphasizing this point because all other libraries I have seen so far confirm to this design. This is important because **a**. we need to comply with other libraries or otherwise there might be state mismatching if candidates are collected during stages before there are changes to candidates (ice-trickle will fix much of this but still important to keep the same.) **b**. This improves resource westfulness. As I mentioned in my personal case I create like 30 connections to other peers. This means if we start allocating candidates with the turn server before connectivity is "imminent" there might be a large amount of resources allocated for no reason. Big waste.

> Maybe change Connect to Start to match aioice and the intent a little better (since we aren't really connecting, but kicking off a complicated stateful process)

Agreed.

> I find RTCDtlsTransport confusing. Especially since it is only used for SCTP traffic, and it shared across all candidates. Currently we share one DTLS state between all ports, and candidates may switch but we always use the same DTLS state. Right now a candidate sends traffic, and just conditionally chooses to encrypt things using DTLS (am I sending SRTP or SCTP).

According to the RFC [HERE](https://www.w3.org/TR/webrtc/#rtcdtlstransport-interface) `RTCDtlsTransport` is part of the specifications. In this current run over the code I am not planning on creating `RTCDtlsTransport` - just to be clear. After the initial implementation of `RTCIceTransport` I was planning on making it backward compatible so we could integrate it first and then we could talk about `RTCDtlsTransport`. I have solid understanding of crypto implementations so happy to take over the task of implementing golang pure dtls. When I do that we could have a deep talk about `RTCDtlsTransport`. Same applies for SCTP. As I said, the goal is going to be backward compatible.

As for the flow, I am not entirely sure how to "backward compatible" will look like but here is the flow of that aioice uses. There are a few of them depended on whether it is audio/video stream or data.
**data**: `createDataChannel()` -> `NewRTCDataChannel(createSctpTransport(), parameters)` -> `NewRTCSctpTransport(createDtlsTransport())` -> `NewRTCIceTransport(configuration.iceServers)`
**stream**: `addTrack()` -> `createTransceiver()` -> `createDtlsTransport()` -> `NewRTCIceTransport(configuration.iceServers)` -> `RTCRtpTransceiver(RTCRtpSender(dtlsTransport), RTCRtpReceiver(dtlsTransport))`

In general some of the design that I omitted is that `IceAgent` also has additional functions:
```
class IceAgent:
    AddRemoteCandidate()
    Gather()
    Connect()
    SetSelectedPair()
    Send()
    SendTo()
    Recv()
    RecvFrom()
```
This way we can then wrap these calls anyway we want in terms of actually sending the data over the connections. I am not entirely sure how exactly the stack will look like but I am sure we can make it anyway you want since the actual sending and receiving code is going to be within the ice and we'll just wrap it around in any kind of stack you wish for.
